src/Objects/Containers/LinkedCell/LinkedCellContainer.cpp

```diff
@@ -0,0 +1,263 @@
+            //SPDLOG_ERROR("x {} y {} z {} {} {} {} {} ", x, y, z, numCellsInXDim, numCellsInYDim, numCellsInZDim, cellSizePerDimension[2]);
+            return -1;
+        }
+
+        const int strideYZ = (numCellsInYDim + 2)*1;
+        const int strideZ = 1; // for 2d
+        return (x + 1) * strideYZ + (y + 1) * strideZ; // for 2d
+    }
+
+
+    Cell *LinkedCellContainer::mapParticleToCell(const Particle &particle) {
+        std::array<double, 3> particlePosition = particle.getX();
+        std::array<int, 3> cellPosition = {
+            static_cast<int>(std::floor(particlePosition[0] / cellSizePerDimension[0])),
+            static_cast<int>(std::floor(particlePosition[1] / cellSizePerDimension[1])),
+            static_cast<int>(std::floor(particlePosition[2] / cellSizePerDimension[2]))
+        };
+        //SPDLOG_INFO("{} {}", particlePosition[0], cellSizePerDimension[0] );
+        int cellInd = cellIndex(cellPosition[0], cellPosition[1], cellPosition[2]);
+        if (cellInd >= (int)cells.size() || cellInd < 0) {
+            SPDLOG_WARN("The given particle does not belong to any cell!");
+            return nullptr;
+        }
+
+        return &cells.at(cellInd);
+    }
+
+    void LinkedCellContainer::initializeCells() {
+        SPDLOG_DEBUG("Initializing cells...");
+        for (int x = -1; x < cellNumPerDimension[0] + 1; ++x) {
+            for (int y = -1; y < cellNumPerDimension[1] + 1 ; ++y) {
+                    if (x < 0 || y < 0 ||  x >= cellNumPerDimension[0] || y >= cellNumPerDimension[1]) {
+                        Cell nCell(Cell::CType::HALO);
+                        cells.push_back(nCell);
+                        haloCells.push_back(cells.back());
+                    } else if (x == 0 || y == 0 || x == cellNumPerDimension[0] - 1 || y == cellNumPerDimension[1] - 1) {
+                        Cell nCell(Cell::CType::BOUNDARY);
+                        cells.push_back(nCell);
+                        boundaryCells.push_back(cells.back());
+                        innerCells.push_back(cells.back());
+                    } else {
+                        Cell nCell(Cell::CType::INNER);
+                        cells.push_back(nCell);
+                        innerCells.push_back(cells.back());
+                    }
+
+            }
+        }
+    }
+
+    void LinkedCellContainer::deleteHaloParticles() {
+        std::unordered_set<Particle *> particlesDelete;
+        for (auto &cell: haloCells) {
+            for (Particle *particle: cell.get().getParticlesInCell()) {
+                particlesDelete.insert(particle);
+            }
+        }
+
+        auto isInDelete = [&](Particle &particle) {
+            return particlesDelete.find(&particle) != particlesDelete.end();
+        };
+
+        particles.erase(std::remove_if(particles.begin(), particles.end(), isInDelete), particles.end());
```

Comment:
from a performance point of view, deleting particles from the vector is not good, instead you could "mark" them deleted and let them just stay. Potentially, you could reuse them when you need to create a new halo particle later on.